Doesn't compiles to aarch64-pc-windows-msvc via Clang

I tried to disable WITH_NEON but it still generates code with neon calls and fails to link. I commented neon code in count_lengths in inftrees.c but its not solving root of the problem

It's so easy to reproduce
```
-DBUILD_SHARED_LIBS=ON
-DCMAKE_C_COMPILER="clang"
-DCMAKE_SYSTEM_NAME="Windows"
-DCMAKE_SYSTEM_PROCESSOR="arm64"
-DCMAKE_C_FLAGS="-target aarch64-pc-windows-msvc"
```
```
FAILED: CMakeFiles/zlib-ng.dir/inftrees.c.obj 
C:\PROGRA~1\LLVM\bin\clang.exe -DARM_CRC32 -DARM_CRC32_INTRIN -DARM_FEATURES -DARM_NEON -DHAVE_ARM_ACLE_H -DHAVE_ATTRIBUTE_ALIGNED -DHAVE_BUILTIN_ASSUME_ALIGNED -DHAVE_VISIBILITY_HIDDEN -DHAVE_VISIBILITY_INTERNAL -DNO_FSEEKO -DWITH_GZFILEOP -DWITH_OPTIM -DZLIBNG_NATIVE_API -DZLIB_DLL -IC:/Users/Nelonn/Projects/zlib-ng/cmake-build-debug-clang-arm64 -IC:/Users/Nelonn/Projects/zlib-ng -target aarch64-pc-windows-msvc -O0 -std=c11 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -g -Xclang -gcodeview -Wall -DZLIB_DEBUG -MD -MT CMakeFiles/zlib-ng.dir/inftrees.c.obj -MF CMakeFiles\zlib-ng.dir\inftrees.c.obj.d -o CMakeFiles/zlib-ng.dir/inftrees.c.obj -c C:/Users/Nelonn/Projects/zlib-ng/inftrees.c
In file included from C:/Users/Nelonn/Projects/zlib-ng/inftrees.c:15:
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:40:28: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]
   40 | static inline uint16x8x4_t vld1q_u16_x4(uint16_t const *a) {
      |                            ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:6643:68: note: expanded from macro 'vld1q_u16_x4'
 6643 | #define vld1q_u16_x4(src) __n128x4_to_uint16x8x4_t(neon_ld1m4_q16((__int16*)(src)))
      |                                                                    ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:512:46: note: expanded from macro '__n128x4_to_uint16x8x4_t'
  512 | #define __n128x4_to_uint16x8x4_t(x)         (x)
      |                                              ^
In file included from C:/Users/Nelonn/Projects/zlib-ng/inftrees.c:15:
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:40:28: error: function cannot return function type 'int (const uint16_t *)' (aka 'int (const unsigned short *)')
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:6643:67: note: expanded from macro 'vld1q_u16_x4'
 6643 | #define vld1q_u16_x4(src) __n128x4_to_uint16x8x4_t(neon_ld1m4_q16((__int16*)(src)))
      |                                                                   ^
In file included from C:/Users/Nelonn/Projects/zlib-ng/inftrees.c:15:
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:40:28: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:6643:68: note: expanded from macro 'vld1q_u16_x4'
 6643 | #define vld1q_u16_x4(src) __n128x4_to_uint16x8x4_t(neon_ld1m4_q16((__int16*)(src)))
      |                                                                    ^
In file included from C:/Users/Nelonn/Projects/zlib-ng/inftrees.c:15:
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:42:28: error: use of undeclared identifier 'a'
   42 |     ret.val[0] = vld1q_u16(a);
      |                            ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:43:28: error: use of undeclared identifier 'a'
   43 |     ret.val[1] = vld1q_u16(a+8);
      |                            ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:44:28: error: use of undeclared identifier 'a'
   44 |     ret.val[2] = vld1q_u16(a+16);
      |                            ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:45:28: error: use of undeclared identifier 'a'
   45 |     ret.val[3] = vld1q_u16(a+24);
      |                            ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:49:28: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]
   49 | static inline uint8x16x4_t vld1q_u8_x4(uint8_t const *a) {
      |                            ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:6645:66: note: expanded from macro 'vld1q_u8_x4'
 6645 | #define vld1q_u8_x4(src) __n128x4_to_uint8x16x4_t(neon_ld1m4_q8((__int8*)(src)))
      |                                                                  ^
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:515:46: note: expanded from macro '__n128x4_to_uint8x16x4_t'
  515 | #define __n128x4_to_uint8x16x4_t(x)         (x)
      |                                              ^
In file included from C:/Users/Nelonn/Projects/zlib-ng/inftrees.c:15:
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:49:28: error: function cannot return function type 'int (const uint8_t *)' (aka 'int (const unsigned char *)')
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:6645:65: note: expanded from macro 'vld1q_u8_x4'
 6645 | #define vld1q_u8_x4(src) __n128x4_to_uint8x16x4_t(neon_ld1m4_q8((__int8*)(src)))
      |                                                                 ^
In file included from C:/Users/Nelonn/Projects/zlib-ng/inftrees.c:15:
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:49:28: warning: omitting the parameter name in a function definition is a C23 extension [-Wc23-extensions]
C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.44.35207\include\arm64_neon.h:6645:66: note: expanded from macro 'vld1q_u8_x4'
 6645 | #define vld1q_u8_x4(src) __n128x4_to_uint8x16x4_t(neon_ld1m4_q8((__int8*)(src)))
      |                                                                  ^
In file included from C:/Users/Nelonn/Projects/zlib-ng/inftrees.c:15:
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:51:27: error: use of undeclared identifier 'a'
   51 |     ret.val[0] = vld1q_u8(a);
      |                           ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:52:27: error: use of undeclared identifier 'a'
   52 |     ret.val[1] = vld1q_u8(a+16);
      |                           ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:53:27: error: use of undeclared identifier 'a'
   53 |     ret.val[2] = vld1q_u8(a+32);
      |                           ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:54:27: error: use of undeclared identifier 'a'
   54 |     ret.val[3] = vld1q_u8(a+48);
      |                           ^
C:/Users/Nelonn/Projects/zlib-ng\arch/arm/neon_intrins.h:58:20: error: type specifier missing, defaults to 'int'; ISO C99 and later do not support implicit int [-Wimplicit-int]
   58 | static inline void vst1q_u16_x4(uint16_t *p, uint16x8x4_t a) {
      |                    ^
...and many more lines
```
Clang is the de facto standard so we using it on all platforms

--- a/.github/workflows/cmake.yml
+++ b/.github/workflows/cmake.yml
@@ -546,6 +546,11 @@ jobs:
             cmake-args: -A ARM64EC
             # Coverage disabled for msvc
 
+          - name: Windows ClangCl ARM64
+            os: windows-11-arm
+            cmake-args: -T ClangCl -A ARM64
+            # Coverage disabled for clangcl
+
           - name: Windows ClangCl Win32
             os: windows-latest
             cmake-args: -T ClangCl -A Win32

--- a/arch/arm/acle_intrins.h
+++ b/arch/arm/acle_intrins.h
@@ -2,7 +2,7 @@
 #define ARM_ACLE_INTRINS_H
 
 #include <stdint.h>
-#ifdef _MSC_VER
+#if defined(_MSC_VER) && !defined(__clang__)
 #  include <intrin.h>
 #elif defined(HAVE_ARM_ACLE_H)
 #  include <arm_acle.h>

--- a/arch/arm/crc32_armv8_pmull_eor3.c
+++ b/arch/arm/crc32_armv8_pmull_eor3.c
@@ -17,7 +17,7 @@
 
 /* Carryless multiply low 64 bits: a[0] * b[0] */
 static inline uint64x2_t clmul_lo(uint64x2_t a, uint64x2_t b) {
-#ifdef _MSC_VER
+#if defined(_MSC_VER) && !defined(__clang__)
     return vreinterpretq_u64_p128(vmull_p64(
         vget_low_p64(vreinterpret_p64_u64(a)),
         vget_low_p64(vreinterpret_p64_u64(b))));
@@ -35,7 +35,7 @@ static inline uint64x2_t clmul_hi(uint64x2_t a, uint64x2_t b) {
 
 /* Carryless multiply of two 32-bit scalars: a * b (returns 64-bit result in 128-bit vector) */
 static inline uint64x2_t clmul_scalar(uint32_t a, uint32_t b) {
-#ifdef _MSC_VER
+#if defined(_MSC_VER) && !defined(__clang__)
     return vreinterpretq_u64_p128(vmull_p64(vdup_n_p64((poly64_t)a), vdup_n_p64((poly64_t)b)));
 #else
     return vreinterpretq_u64_p128(vmull_p64((poly64_t)a, (poly64_t)b));

--- a/arch/arm/neon_intrins.h
+++ b/arch/arm/neon_intrins.h
@@ -1,7 +1,7 @@
 #ifndef ARM_NEON_INTRINS_H
 #define ARM_NEON_INTRINS_H
 
-#if defined(_MSC_VER) && defined(ARCH_ARM) && defined(ARCH_64BIT)
+#if defined(_MSC_VER) && !defined(__clang__) && defined(ARCH_ARM) && defined(ARCH_64BIT)
 /* arm64_neon.h is MSVC specific */
 #  include <arm64_neon.h>
 #else
@@ -63,12 +63,12 @@ static inline void vst1q_u16_x4(uint16_t *p, uint16x8x4_t a) {
 }
 #  endif // HASLD4 check
 
-#  ifndef _MSC_VER
+#  if !defined(_MSC_VER) || defined(__clang__)
 #    define vld1_u8_ex(p, align) vld1_u8(HINT_ALIGNED((p), (align)/8))
 #    define vld1q_u8_ex(p, align) vld1q_u8(HINT_ALIGNED((p), (align)/8))
 #    define vld1q_u64_ex(p, align) vld1q_u64(HINT_ALIGNED((p), (align)/8))
 #  endif
-#  if !defined(_MSC_VER) || !defined(ARM_NEON_HASLD4)
+#  if !defined(_MSC_VER) || !defined(ARM_NEON_HASLD4) || defined(__clang__)
 #    define vld1q_u8_x4_ex(p, align) vld1q_u8_x4(HINT_ALIGNED((p), (align)/8))
 #    define vld1q_u16_x4_ex(p, align) vld1q_u16_x4(HINT_ALIGNED((p), (align)/8))
 #    define vst1q_u16_x4_ex(p, a, align) vst1q_u16_x4(HINT_ALIGNED((p), (align)/8), a)

--- a/cmake/detect-intrinsics.cmake
+++ b/cmake/detect-intrinsics.cmake
@@ -42,7 +42,7 @@ macro(check_armv8_compiler_flag)
     )
     # Check whether compiler supports ARMv8 intrinsics
     check_c_source_compiles(
-        "#if defined(_MSC_VER)
+        "#if defined(_MSC_VER) && !defined(__clang__)
         #include <intrin.h>
         #else
         #include <arm_acle.h>
@@ -73,12 +73,12 @@ macro(check_armv8_pmull_eor3_compiler_flag)
     # Check whether compiler supports ARMv8 PMULL + EOR3 intrinsics
     set(CMAKE_REQUIRED_FLAGS "${PMULLEOR3FLAG} ${NATIVEFLAG} ${ZNOLTOFLAG}")
     check_c_source_compiles(
-        "#if defined(_MSC_VER) && (defined(_M_ARM64) || defined(_M_ARM64EC))
+        "#if defined(_MSC_VER) && !defined(__clang__) && (defined(_M_ARM64) || defined(_M_ARM64EC))
         #  include <arm64_neon.h>
         #else
         #  include <arm_neon.h>
         #endif
-        #ifdef _MSC_VER
+        #if defined(_MSC_VER) && !defined(__clang__)
         __n128 f(__n64 a, __n64 b) {
         #else
         poly128_t f(poly64_t a, poly64_t b) {
@@ -260,7 +260,7 @@ macro(check_neon_compiler_flag)
     # Check whether compiler supports NEON flag
     set(CMAKE_REQUIRED_FLAGS "${NEONFLAG} ${NATIVEFLAG} ${ZNOLTOFLAG}")
     check_c_source_compiles(
-        "#if defined(_M_ARM64) || defined(_M_ARM64EC)
+        "#if defined(_MSC_VER) && !defined(__clang__) && (defined(_M_ARM64) || defined(_M_ARM64EC))
         #  include <arm64_neon.h>
         #else
         #  include <arm_neon.h>
@@ -323,7 +323,7 @@ macro(check_neon_ld4_intrinsics)
     # Check whether compiler supports loading 4 neon vecs into a register range
     set(CMAKE_REQUIRED_FLAGS "${NEONFLAG} ${NATIVEFLAG} ${ZNOLTOFLAG}")
     check_c_source_compiles(
-        "#if defined(_MSC_VER) && (defined(_M_ARM64) || defined(_M_ARM64EC))
+        "#if defined(_MSC_VER) && !defined(__clang__) && (defined(_M_ARM64) || defined(_M_ARM64EC))
         #  include <arm64_neon.h>
         #else
         #  include <arm_neon.h>

